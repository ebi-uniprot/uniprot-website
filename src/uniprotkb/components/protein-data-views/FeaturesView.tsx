import React, { useCallback, useMemo } from 'react';
import { html, TemplateResult } from 'lit-html';
import { v1 } from 'uuid';
import { Loader } from 'franklin-sites';
import { uniq } from 'lodash-es';

import useCustomElement from '../../../shared/hooks/useCustomElement';

import { Evidence } from '../../types/modelTypes';
import FeatureType from '../../types/featureType';
import { UniProtProtvistaEvidenceTag } from './UniProtKBEvidenceTag';
import FeaturesTableView, { FeaturesTableCallback } from './FeaturesTableView';
import { Xref } from '../../types/commentTypes';

export enum LocationModifier {
  EXACT = 'EXACT',
  OUTSIDE = 'OUTSIDE',
  UNSURE = 'UNSURE',
  UNKNOWN = 'UNKNOWN',
}

type FeatureLocation = {
  value: number;
  modifier: LocationModifier;
};

export type FeatureData = {
  type: FeatureType;
  featureId?: string;
  description?: string;
  location: {
    start: FeatureLocation;
    end: FeatureLocation;
    sequence?: string;
  };
  alternativeSequence?: {
    originalSequence?: string;
    alternativeSequences: string[];
  };
  evidences?: Evidence[];
  featureCrossReference?: Xref;
}[];

export type ProtvistaFeature = {
  type: string;
  description: string;
  evidences: Evidence[];
  start: number;
  end: number;
  startModifier: LocationModifier;
  endModifier: LocationModifier;
};

type FeatureProps = {
  sequence?: string;
  features: FeatureData;
};

type Fragment = {
  start: number;
  end: number;
  shape?: string;
};

export type ProcessedFeature = {
  protvistaFeatureId: string;
  featureId?: string;
  start: number;
  end: number;
  startModifier: LocationModifier;
  endModifier: LocationModifier;
  type: FeatureType;
  description?: string;
  evidences?: Evidence[];
  sequence?: string;
  locations?: { fragments: Fragment[] }[];
};

export const processFeaturesData = (
  data: FeatureData,
  sequence?: string
): ProcessedFeature[] =>
  data.map(
    (feature): ProcessedFeature => ({
      protvistaFeatureId: feature.featureId || v1(),
      featureId: feature.featureId,
      start: feature.location.start.value,
      end: feature.location.end.value,
      startModifier: feature.location.start.modifier,
      endModifier: feature.location.end.modifier,
      type: feature.type,
      description: feature.description,
      evidences: feature.evidences,
      sequence: sequence?.substring(
        feature.location.start.value - 1,
        feature.location.end.value
      ),
    })
  );

const FeaturesView: React.FC<FeatureProps> = ({
  sequence,
  features,
}): JSX.Element | null => {
  const navigationDefined = useCustomElement(
    /* istanbul ignore next */
    () =>
      import(
        /* webpackChunkName: "protvista-navigation" */ 'protvista-navigation'
      ),
    'protvista-navigation'
  );
  const sequenceDefined = useCustomElement(
    /* istanbul ignore next */
    () =>
      import(/* webpackChunkName: "protvista-sequence" */ 'protvista-sequence'),
    'protvista-sequence'
  );
  const managerDefined = useCustomElement(
    /* istanbul ignore next */
    () =>
      import(/* webpackChunkName: "protvista-manager" */ 'protvista-manager'),
    'protvista-manager'
  );
  const ceDefined = navigationDefined && sequenceDefined && managerDefined;

  const processedData = useMemo(() => processFeaturesData(features, sequence), [
    features,
    sequence,
  ]);
  const featureTypes = useMemo(
    () => uniq(features.map(({ type }) => type.toLowerCase())),
    [features]
  );

  const getColumnConfig = (evidenceTagCallback: FeaturesTableCallback) => ({
    type: {
      label: 'Type',
      resolver: (d: ProcessedFeature): string => d.type,
    },
    id: {
      label: 'ID',
      resolver: (d: ProcessedFeature): string => d.featureId || '',
    },
    positions: {
      label: 'Positions',
      resolver: (d: ProcessedFeature): string =>
        `${d.startModifier === LocationModifier.UNKNOWN ? '?' : d.start}-${
          d.endModifier === LocationModifier.UNKNOWN ? '?' : d.end
        }`,
    },
    description: {
      label: 'Description',
      resolver: (d: ProcessedFeature): TemplateResult =>
        html`
          ${d.description}
          ${d.evidences &&
          UniProtProtvistaEvidenceTag(d.evidences, evidenceTagCallback)}
        `,
    },
    sequence: {
      label: 'Sequence',
      child: true,
      resolver: (d: ProcessedFeature) => d?.sequence || '',
    },
  });

  const trackDefined = useCustomElement(
    /* istanbul ignore next */
    () => import(/* webpackChunkName: "protvista-track" */ 'protvista-track'),
    'protvista-track'
  );

  const setTrackData = useCallback(
    (node): void => {
      if (node && trackDefined) {
        // eslint-disable-next-line no-param-reassign
        node.data = processedData;
      }
    },
    [trackDefined, processedData]
  );

  if (processedData.length === 0) {
    return null;
  }

  if (!ceDefined) {
    return <Loader />;
  }

  return (
    <>
      <h3>Features</h3>
      <p>Showing features for {featureTypes.join(', ')}.</p>
      <protvista-manager attributes="highlight displaystart displayend selectedid">
        {sequence && (
          <>
            <protvista-navigation length={sequence.length} />
            <protvista-track
              ref={setTrackData}
              length={sequence.length}
              layout="non-overlapping"
            />
            <protvista-sequence
              sequence={sequence}
              length={sequence.length}
              height="20"
            />
          </>
        )}
        <FeaturesTableView
          data={processedData}
          getColumnConfig={getColumnConfig}
        />
      </protvista-manager>
    </>
  );
};

export default FeaturesView;
