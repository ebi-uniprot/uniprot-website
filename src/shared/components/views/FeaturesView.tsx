import {
  Fragment,
  lazy,
  ReactNode,
  useCallback,
  useMemo,
  useState,
} from 'react';
import { Feature } from '@nightingale-elements/nightingale-track';

import LazyComponent from '../LazyComponent';

import { useSmallScreen } from '../../hooks/useMatchMedia';

import FeatureTypeHelpMappings from '../../../help/config/featureTypeHelpMappings';

import FeatureType from '../../../uniprotkb/types/featureType';
import { UniParcProcessedFeature } from '../../../uniparc/components/entry/UniParcFeaturesView';
import { Evidence } from '../../../uniprotkb/types/modelTypes';
import { ConfidenceScore } from '../../../uniprotkb/components/protein-data-views/UniProtKBFeaturesView';
import {
  Ligand,
  LigandPart,
} from '../../../uniprotkb/components/protein-data-views/LigandDescriptionView';
import TableFromData from '../table/TableFromData';

const VisualFeaturesView = lazy(
  () =>
    import(/* webpackChunkName: "visual-feature-view" */ './VisualFeaturesView')
);

export type Fragment = {
  start: number;
  end: number;
  shape?: string;
};

export type LocationModifier = 'EXACT' | 'OUTSIDE' | 'UNSURE' | 'UNKNOWN';

/*
export type Feature = {
  accession: string;
  color?: string;
  fill?: string;
  shape?: Shapes;
  tooltipContent?: string;
  type?: string;
  locations?: Array<FeatureLocation>;
  feature?: Feature;
  start?: number;
  end?: number;
  opacity?: number;

  export type VariationDatum = {
  accession: string;
  variant: string;
  start: number;
  size?: number;
  xrefNames: string[];
  hasPredictions: boolean;
  tooltipContent?: string;
  alternativeSequence?: string;
  internalId?: string;
  wildType?: string;
  color?: string;
  consequenceType: string;
};
};

*/
export type ProcessedFeature = Feature & {
  start: number;
  end: number;
  startModifier?: LocationModifier;
  endModifier?: LocationModifier;
  type: FeatureType;
  description?: ReactNode;
  evidences?: Evidence[];
  sequence?: string;
  source?: string;
  primaryAccession?: string;
  // PTM specific
  confidenceScore?: ConfidenceScore;
  // Binding site
  ligand?: Ligand;
  ligandPart?: LigandPart;
  ligandDescription?: string;
};

export type GenericFeature =
  | ProcessedFeature
  // | TransformedVariant // TODO: do we need this here?
  | UniParcProcessedFeature;

type FeatureProps = {
  features: GenericFeature[];
  table: JSX.Element;
  trackHeight?: number;
  sequence?: string;
  withTitle?: boolean;
  noLinkToFullView?: boolean;
};

const FeaturesView = ({
  sequence,
  features,
  rowExtraContent,
  columns,
  trackHeight,
  withTitle = true,
  noLinkToFullView,
}: FeatureProps) => {
  const isSmallScreen = useSmallScreen();
  const [highlightedFeature, setHighlightedFeature] = useState();

  const featureTypes = useMemo(
    () => Array.from(new Set<FeatureType>(features.map(({ type }) => type))),
    [features]
  );

  const handleRowClick = useCallback((datum) => {
    console.log('Clicked feature:', datum);
    setHighlightedFeature(datum.accession);
  }, []);

  const handleFeatureClick = useCallback((feature) => {
    console.log('Clicked feature:', feature);
    setHighlightedFeature(feature.accession);
  }, []);

  return features.length === 0 ? null : (
    <>
      {withTitle && (
        <>
          <h3>Features</h3>
          <p>
            Showing features for{' '}
            {featureTypes.map((featureType, i) => (
              <Fragment key={featureType}>
                {i > 0 && ', '}
                {featureType === 'Other' ? (
                  'other'
                ) : (
                  <span data-article-id={FeatureTypeHelpMappings[featureType]}>
                    {featureType.toLowerCase()}
                  </span>
                )}
              </Fragment>
            ))}
            .
          </p>
        </>
      )}
      {sequence && (
        <LazyComponent
          render={isSmallScreen ? false : undefined}
          fallback={null}
        >
          <VisualFeaturesView
            features={features}
            sequence={sequence}
            trackHeight={trackHeight}
            noLinkToFullView={noLinkToFullView}
            onFeatureClick={handleFeatureClick}
          />
        </LazyComponent>
      )}
      <TableFromData
        data={features}
        columns={columns}
        rowExtraContent={rowExtraContent}
        onRowClick={handleRowClick}
        highlightedFeature={highlightedFeature}
      />
    </>
  );
};

export default FeaturesView;
