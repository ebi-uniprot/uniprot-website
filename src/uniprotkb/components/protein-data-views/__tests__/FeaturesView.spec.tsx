import { render } from '@testing-library/react';

import FeaturesView from '../UniProtKBFeaturesView';

import FeaturesUIData from './__mocks__/featuresUIData';

describe('FeaturesView component', () => {
  test('it renders without crashing', () => {
    const { asFragment } = render(
      <FeaturesView
        primaryAccession="P05067"
        features={FeaturesUIData}
        sequence="ASDASDASASDASDASDSASD"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
