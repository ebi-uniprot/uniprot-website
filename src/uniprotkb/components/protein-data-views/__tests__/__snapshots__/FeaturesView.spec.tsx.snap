// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturesView component it renders without crashing 1`] = `
<DocumentFragment>
  <h3>
    Features
  </h3>
  <p>
    Showing features for 
    <span
      data-article-id="signal"
    >
      signal
    </span>
    .
  </p>
  <protvista-manager
    attributes="highlight displaystart displayend selectedid"
  >
    <div
      class="datatable-with-toggle"
    >
      <protvista-datatable
        filter-scroll="true"
      >
        <table>
          <thead>
            <tr>
              <th
                data-filter="type"
              >
                Type
              </th>
              <th>
                ID
              </th>
              <th>
                Position(s)
              </th>
              <th>
                Description
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              data-end="17"
              data-id="0"
              data-start="1"
            >
              <td
                data-filter="type"
                data-filter-value="Signal"
              >
                Signal
              </td>
              <td />
              <td>
                1-17
              </td>
              <td
                id="description1-17"
              />
              <td>
                <div
                  class="button-group"
                >
                  <a
                    class="button tertiary"
                    href="/blast?ids=P05067[1-17]"
                    title="BLAST the sequence corresponding to this feature"
                  >
                    BLAST
                  </a>
                  <button
                    class="button tertiary no-small"
                    title="Add 1 entry to the basket"
                    type="button"
                  >
                    <test-file-stub />
                    Add
                  </button>
                </div>
              </td>
            </tr>
            <tr
              data-end="17"
              data-group-for="0"
              data-start="1"
            >
              <td>
                <strong>
                  Sequence: 
                </strong>
                ASDASDASASDASDASD
              </td>
            </tr>
          </tbody>
        </table>
      </protvista-datatable>
      <button
        class="button primary toggle-button"
        type="button"
      >
        Expand table
      </button>
    </div>
  </protvista-manager>
</DocumentFragment>
`;
